From 89ad3965867892f15059a323faf685d3c3fe6e15 Mon Sep 17 00:00:00 2001
From: Thierry Vignaud <thierry.vignaud@gmail.com>
Date: Fri, 21 Oct 2016 11:06:43 +0200
Subject: [PATCH] fix testsuite (adjust pkg list)

---
 tests/rpmpython.at | 1 +
 1 file changed, 1 insertion(+)

diff --git a/tests/rpmpython.at b/tests/rpmpython.at
index 949673b..a6405fd 100644
--- a/tests/rpmpython.at
+++ b/tests/rpmpython.at
@@ -79,6 +79,7 @@ myprint(spec.sourceHeader.format('%{nvr}'))
 [src hello-1.0-modernize.patch 0 2
 src hello-1.0.tar.gz 0 1
 hello-1.0-1
+hello-debuginfo-1.0-1
 hello-1.0-1
 ])
 
-- 
2.10.1

