From 6e23e20263ce093289d81c61233833b9013343c9 Mon Sep 17 00:00:00 2001
From: Lubos Kardos <lkardos@redhat.com>
Date: Thu, 28 Jan 2016 09:06:17 +0100
Subject: [PATCH 01/18] Fix memory leak in file triggers

---
 lib/rpmtriggers.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/lib/rpmtriggers.c b/lib/rpmtriggers.c
index 4c6e821..1f97623 100644
--- a/lib/rpmtriggers.c
+++ b/lib/rpmtriggers.c
@@ -610,6 +610,7 @@ rpmRC runImmedFileTriggers(rpmts ts, rpmte te, rpmsenseFlags sense,
 					    triggers->triggerInfo[i].tix);
     }
     rpmtriggersFree(triggers);
+    headerFree(trigH);
 
     return (nerrors == 0) ? RPMRC_OK : RPMRC_FAIL;
 }
-- 
2.7.0

