From 4d65d0139464386eaa56419585442713dd43726d Mon Sep 17 00:00:00 2001
From: Thierry Vignaud <thierry.vignaud@gmail.com>
Date: Fri, 21 Oct 2016 10:30:22 +0200
Subject: [PATCH] fix calling sepdebugcrcfix

---
 scripts/find-debuginfo.sh | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/scripts/find-debuginfo.sh b/scripts/find-debuginfo.sh
index 6aaf216..bb5d7c2 100644
--- a/scripts/find-debuginfo.sh
+++ b/scripts/find-debuginfo.sh
@@ -371,7 +371,7 @@ if $run_dwz && type dwz >/dev/null 2>&1 \
     # dwz invalidates .gnu_debuglink CRC32 in the main files.
     cat "$ELFBINSFILE" |
     (cd "$RPM_BUILD_ROOT"; \
-     xargs -d '\n' ${lib_rpm_dir}/sepdebugcrcfix usr/lib/debug)
+     xargs -d '\n' /usr/lib/rpm/sepdebugcrcfix usr/lib/debug)
   fi
 fi
 
-- 
2.10.1

