From 8efe51e8c24b7739f0bf7680e21083c8964633f5 Mon Sep 17 00:00:00 2001
From: Florian Festi <ffesti@redhat.com>
Date: Fri, 15 Jan 2016 14:32:47 +0100
Subject: [PATCH] Add support for %missingok as a standalone file attribute

---
 build/files.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/build/files.c b/build/files.c
index d00a785..2d72633 100644
--- a/build/files.c
+++ b/build/files.c
@@ -783,6 +783,7 @@ static VFA_t const virtualAttrs[] = {
     { "%readme",	RPMFILE_README },
     { "%license",	RPMFILE_LICENSE },
     { "%pubkey",	RPMFILE_PUBKEY },
+    { "%missingok",	RPMFILE_MISSINGOK },
     { NULL, 0 }
 };
 
-- 
2.1.0

