From 454589a7dae68d58caebee834e8251075728dd26 Mon Sep 17 00:00:00 2001
From: unknown author <cooker@mandrivalinux.org>
Date: Mon, 5 Jan 2009 13:29:57 +0000
Subject: [PATCH 12/35] revert unused skipDir functionality

---
 lib/rpmdb.c |   28 ----------------------------
 1 files changed, 0 insertions(+), 28 deletions(-)

diff --git a/lib/rpmdb.c b/lib/rpmdb.c
index 913b870..fd2b3c5 100644
--- a/lib/rpmdb.c
+++ b/lib/rpmdb.c
@@ -2920,31 +2920,6 @@ exit:
     return ret;
 }
 
-#define _skip(_dn)	{ sizeof(_dn)-1, (_dn) }
-
-static struct skipDir_s {
-    int dnlen;
-    const char * dn;
-} skipDirs[] = {
-    { 0, NULL }
-};
-
-static int skipDir(const char * dn)
-{
-    struct skipDir_s * sd = skipDirs;
-    int dnlen;
-
-    dnlen = strlen(dn);
-    for (sd = skipDirs; sd->dn != NULL; sd++) {
-	if (dnlen < sd->dnlen)
-	    continue;
-	if (strncmp(dn, sd->dn, sd->dnlen))
-	    continue;
-	return 1;
-    }
-    return 0;
-}
-
 /* XXX transaction.c */
 int rpmdbFindFpList(rpmdb db, fingerPrint * fpList, dbiIndexSet * matchList, 
 		    int numItems)
@@ -2975,9 +2950,6 @@ int rpmdbFindFpList(rpmdb db, fingerPrint * fpList, dbiIndexSet * matchList,
 	if (key->size == 0) 
 	    key->size++;	/* XXX "/" fixup. */
 
-	if (skipDir(fpList[i].entry->dirName))
-	    continue;
-
 	xx = rpmdbGrowIterator(mi, i);
 
     }
-- 
1.6.1.3

