From 237387f1423ed3a1db7ea8f98edf5ea1da5c6ec8 Mon Sep 17 00:00:00 2001
From: unknown author <cooker@mandrivalinux.org>
Date: Mon, 5 Jan 2009 13:29:57 +0000
Subject: [PATCH 02/35] non pre scripts dont fail

---
 lib/psm.c |   23 ++++++++---------------
 1 files changed, 8 insertions(+), 15 deletions(-)

diff --git a/lib/psm.c b/lib/psm.c
index 9070c6d..890b223 100644
--- a/lib/psm.c
+++ b/lib/psm.c
@@ -1398,16 +1398,14 @@ assert(psm->mi == NULL);
 
 	    if (!(rpmtsFlags(ts) & RPMTRANS_FLAG_NOTRIGGERUN)) {
 		/* Run triggers in this package other package(s) set off. */
-		rc = rpmpsmNext(psm, PSM_IMMED_TRIGGERS);
-		if (rc) break;
+		rpmpsmNext(psm, PSM_IMMED_TRIGGERS);
 
 		/* Run triggers in other package(s) this package sets off. */
-		rc = rpmpsmNext(psm, PSM_TRIGGERS);
-		if (rc) break;
+		rpmpsmNext(psm, PSM_TRIGGERS);
 	    }
 
 	    if (!(rpmtsFlags(ts) & RPMTRANS_FLAG_NOPREUN))
-		rc = rpmpsmNext(psm, PSM_SCRIPT);
+		rpmpsmNext(psm, PSM_SCRIPT);
 	}
 	break;
     case PSM_PROCESS:
@@ -1542,17 +1540,14 @@ assert(psm->mi == NULL);
 	    psm->countCorrection = 0;
 
 	    if (!(rpmtsFlags(ts) & RPMTRANS_FLAG_NOPOST)) {
-		rc = rpmpsmNext(psm, PSM_SCRIPT);
-		if (rc) break;
+		rpmpsmNext(psm, PSM_SCRIPT);
 	    }
 	    if (!(rpmtsFlags(ts) & RPMTRANS_FLAG_NOTRIGGERIN)) {
 		/* Run triggers in other package(s) this package sets off. */
-		rc = rpmpsmNext(psm, PSM_TRIGGERS);
-		if (rc) break;
+		rpmpsmNext(psm, PSM_TRIGGERS);
 
 		/* Run triggers in this package other package(s) set off. */
-		rc = rpmpsmNext(psm, PSM_IMMED_TRIGGERS);
-		if (rc) break;
+		rpmpsmNext(psm, PSM_IMMED_TRIGGERS);
 	    }
 
 	    if (!(rpmtsFlags(ts) & RPMTRANS_FLAG_APPLYONLY))
@@ -1567,14 +1562,12 @@ assert(psm->mi == NULL);
 	    psm->countCorrection = -1;
 
 	    if (!(rpmtsFlags(ts) & RPMTRANS_FLAG_NOPOSTUN)) {
-		rc = rpmpsmNext(psm, PSM_SCRIPT);
-		if (rc) break;
+		rpmpsmNext(psm, PSM_SCRIPT);
 	    }
 
 	    if (!(rpmtsFlags(ts) & RPMTRANS_FLAG_NOTRIGGERPOSTUN)) {
 		/* Run triggers in other package(s) this package sets off. */
-		rc = rpmpsmNext(psm, PSM_TRIGGERS);
-		if (rc) break;
+		rpmpsmNext(psm, PSM_TRIGGERS);
 	    }
 
 	    if (!(rpmtsFlags(ts) & RPMTRANS_FLAG_APPLYONLY))
-- 
1.6.1.3

