From 0bf391f94e85851d35a7adb4f1fcff17253825fe Mon Sep 17 00:00:00 2001
From: unknown author <cooker@mandrivalinux.org>
Date: Mon, 5 Jan 2009 13:29:57 +0000
Subject: [PATCH 14/35] allow conflicting ghost files

---
 lib/rpmfi.c |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/lib/rpmfi.c b/lib/rpmfi.c
index 6055a24..76f6626 100644
--- a/lib/rpmfi.c
+++ b/lib/rpmfi.c
@@ -472,6 +472,9 @@ int rpmfiCompare(const rpmfi afi, const rpmfi bfi)
     if ((rpmfiFFlags(afi) & RPMFILE_GHOST) ||
 	(rpmfiFFlags(bfi) & RPMFILE_GHOST)) return 0;
 
+    if ((rpmfiFFlags(afi) & RPMFILE_GHOST) ||
+	(rpmfiFFlags(bfi) & RPMFILE_GHOST)) return 0;
+
     if (awhat != bwhat) return 1;
 
     if (awhat == LINK) {
-- 
1.6.1.3

