From 117973f303e57ee92a9aaa5577ca9c540c33c3bf Mon Sep 17 00:00:00 2001
From: Colin Guthrie <colin@mageia.org>
Date: Wed, 23 Jul 2014 00:10:02 +0100
Subject: [PATCH] filetriggers: Fix various warnings.

---
 lib/filetriggers.c | 15 ++++++++++++---
 1 file changed, 12 insertions(+), 3 deletions(-)

diff --git a/lib/filetriggers.c b/lib/filetriggers.c
index 8fbc751..589007c 100644
--- a/lib/filetriggers.c
+++ b/lib/filetriggers.c
@@ -1,4 +1,8 @@
 #include "system.h"
+
+#include <sys/types.h>
+#include <sys/wait.h>
+
 #include <rpm/rpmlib.h>
 #include <rpm/rpmlog.h>
 #include <rpm/rpmmacro.h>	/* XXX for rpmExpand */
@@ -207,7 +211,10 @@ static int popen_with_root(const char *rootDir, const char *cmd, int *pid)
 		    rpmlog(RPMLOG_ERR, "chroot to %s failed\n", rootDir);
 		    _exit(-1);
 	       }
-	       chdir("/");
+	       if (chdir("/") != 0) {
+		    rpmlog(RPMLOG_ERR, "chdir to / (inside chroot) failed\n");
+		    _exit(-1);
+	       }
 	  }
 	  const char *argv[2];
 	  argv[0] = cmd;
-- 
1.9.2

