diff -p -up rpm-4.6.0/lib/rpmtag.h.rpm5_tags~ rpm-4.6.0/lib/rpmtag.h
--- rpm-4.6.0/lib/rpmtag.h.rpm5_tags~	2009-05-22 11:28:23.605844290 -0400
+++ rpm-4.6.0/lib/rpmtag.h	2009-05-22 11:28:29.496942572 -0400
@@ -259,7 +259,7 @@ typedef enum rpmTag_e {
     RPMTAG_REPOTAG		= 1181,	/* s (unimplemented) */
     RPMTAG_KEYWORDS		= 1182,	/* s[] (unimplemented) */
     RPMTAG_BUILDPLATFORMS	= 1183,	/* s[] (unimplemented) */
-    RPMTAG_PACKAGECOLOR		= 1184, /* i (unimplemented) */
+    RPMTAG_PACKAGECOLOR		= 1184, /* i */
     RPMTAG_PACKAGEPREFCOLOR	= 1185, /* i (unimplemented) */
     RPMTAG_XATTRSDICT		= 1186, /* s[] (unimplemented) */
     RPMTAG_FILEXATTRSX		= 1187, /* i[] (unimplemented) */
@@ -270,6 +270,31 @@ typedef enum rpmTag_e {
     RPMTAG_REQUIREATTRSX	= 1192, /* i[] (unimplemented) */
     RPMTAG_BUILDPROVIDES	= 1193, /* internal */
     RPMTAG_BUILDOBSOLETES	= 1194, /* internal */
+    RPMTAG_DBINSTANCE		= 1195, /* i */
+    RPMTAG_NVRA			= 1196, /* s */
+    RPMTAG_FILEPATHS		= 1197, /* s[] */
+    RPMTAG_ORIGPATHS		= 1198, /* s[] */
+    RPMTAG_RPMLIBVERSION	= 1199, /* i */
+    RPMTAG_RPMLIBTIMESTAMP	= 1200, /* i */
+    RPMTAG_RPMLIBVENDOR		= 1201, /* i */
+    RPMTAG_CLASS		= 1202, /* s arbitrary */
+    RPMTAG_TRACK		= 1203, /* s internal arbitrary */
+    RPMTAG_TRACKPROG		= 1204, /* s internal arbitrary */
+    RPMTAG_SANITYCHECK		= 1205, /* s */
+    RPMTAG_SANITYCHECKPROG	= 1206, /* s */
+    RPMTAG_FILESTAT		= 1207, /* s[] stat(2) from metadata extension*/
+    RPMTAG_STAT			= 1208, /* s[] stat(2) from disk extension */
+    RPMTAG_ORIGINTID		= 1209,	/* i[] */
+    RPMTAG_ORIGINTIME		= 1210,	/* i[] */
+    RPMTAG_HEADERSTARTOFF	= 1211,	/* l */
+    RPMTAG_HEADERENDOFF		= 1212,	/* l */
+    RPMTAG_PACKAGETIME		= 1213,	/* l */
+    RPMTAG_PACKAGESIZE		= 1214,	/* l */
+    RPMTAG_PACKAGEDIGEST	= 1215,	/* s */
+    RPMTAG_PACKAGESTAT		= 1216,	/* x */
+    RPMTAG_PACKAGEBASEURL	= 1217,	/* s */
+    RPMTAG_DISTEPOCH		= 1218, /* s */
+
     RPMTAG_FILENAMES		= 5000, /* s[] extension */
     RPMTAG_FILEPROVIDE		= 5001, /* s[] extension */
     RPMTAG_FILEREQUIRE		= 5002, /* s[] extension */
