From cb421f706acd23f6b2c06a0cf3357ef49c4f7041 Mon Sep 17 00:00:00 2001
From: unknown author <cooker@mandrivalinux.org>
Date: Mon, 5 Jan 2009 13:29:57 +0000
Subject: [PATCH 10/36] check file trim double slash in buildroot

---
 scripts/check-files |    6 ++++--
 1 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/scripts/check-files b/scripts/check-files
index b397f1c..31e92fe 100755
--- a/scripts/check-files
+++ b/scripts/check-files
@@ -18,6 +18,7 @@ if [ ! -d "${RPM_BUILD_ROOT}" ] ; then
 	exit 1
 fi
 
+RPM_BUILD_ROOT=`echo $RPM_BUILD_ROOT | sed 's://*:/:g'`
 # Create temporary file listing files in the manifest
 [ -n "$TMPDIR" ] || TMPDIR="/tmp"
 FILES_DISK=`mktemp "${TMPDIR}/rpmXXXXXX"`
@@ -27,6 +28,6 @@ trap "rm -f \"${FILES_DISK}\"" 0 2 3 5 1
 
 # Find non-directory files in the build root and compare to the manifest.
 # TODO: regex chars in last sed(1) expression should be escaped
-find "${RPM_BUILD_ROOT}" -type f -o -type l | LC_ALL=C sort > "${FILES_DISK}"
-LC_ALL=C sort | diff -d "${FILES_DISK}" - | sed -n 's|^< '"${RPM_BUILD_ROOT}"'\(.*\)$|   \1|gp'
+find "$RPM_BUILD_ROOT" -type f -o -type l | sed 's://*:/:g' | LC_ALL=C sort > "${FILES_DISK}"
+sed 's://*:/:g' | LC_ALL=C sort -u | diff -d "${FILES_DISK}" - | sed -n 's|^< '"${RPM_BUILD_ROOT}"'\(.*\)$|   \1|gp'
 
