--- a/lib/psm.c
+++ b/lib/psm.c	2011-06-26 22:14:37.258432871 +0200
@@ -694,8 +694,7 @@
 	if (rc) break;
 
 	if (!(rpmtsFlags(ts) & RPMTRANS_FLAG_NOPOST)) {
-	    rc = runInstScript(psm, RPMTAG_POSTIN);
-	    if (rc) break;
+	    runInstScript(psm, RPMTAG_POSTIN);
 	}
 	if (!(rpmtsFlags(ts) & RPMTRANS_FLAG_NOTRIGGERIN)) {
 	    /* Run triggers in other package(s) this package sets off. */
@@ -746,16 +745,14 @@
 
 	if (!(rpmtsFlags(ts) & RPMTRANS_FLAG_NOTRIGGERUN)) {
 	    /* Run triggers in this package other package(s) set off. */
-	    rc = runImmedTriggers(psm, RPMSENSE_TRIGGERUN);
-	    if (rc) break;
+	    runImmedTriggers(psm, RPMSENSE_TRIGGERUN);
 
 	    /* Run triggers in other package(s) this package sets off. */
-	    rc = runTriggers(psm, RPMSENSE_TRIGGERUN);
-	    if (rc) break;
+	    runTriggers(psm, RPMSENSE_TRIGGERUN);
 	}
 
 	if (!(rpmtsFlags(ts) & RPMTRANS_FLAG_NOPREUN))
-	    rc = runInstScript(psm, RPMTAG_PREUN);
+	    runInstScript(psm, RPMTAG_PREUN);
 
 	/* Run file triggers in this package other package(s) set off. */
 	rc = runImmedFileTriggers(psm->ts, psm->te, RPMSENSE_TRIGGERUN,
@@ -775,14 +772,12 @@
 			    RPMSCRIPT_FILETRIGGER, 1);
 
 	if (!(rpmtsFlags(ts) & RPMTRANS_FLAG_NOPOSTUN)) {
-	    rc = runInstScript(psm, RPMTAG_POSTUN);
-	    if (rc) break;
+	    runInstScript(psm, RPMTAG_POSTUN);
 	}
 
 	if (!(rpmtsFlags(ts) & RPMTRANS_FLAG_NOTRIGGERPOSTUN)) {
 	    /* Run triggers in other package(s) this package sets off. */
-	    rc = runTriggers(psm, RPMSENSE_TRIGGERPOSTUN);
-	    if (rc) break;
+	    runTriggers(psm, RPMSENSE_TRIGGERPOSTUN);
 	}
 
 	/* Run file triggers in other package(s) this package sets off. */
