From 1b6ce73e13bf84a5984d3a37608f043f919a5fe2 Mon Sep 17 00:00:00 2001
From: unknown author <cooker@mandrivalinux.org>
Date: Mon, 5 Jan 2009 13:29:57 +0000
Subject: [PATCH 28/36] copyright tag

---
 build/parsePreamble.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/build/parsePreamble.c b/build/parsePreamble.c
index 9751cdd..40e7c39 100644
--- a/build/parsePreamble.c
+++ b/build/parsePreamble.c
@@ -729,6 +728,7 @@ static struct PreambleRec_s const preambleList[] = {
     {RPMTAG_EPOCH,		0, 0, LEN_AND_STR("epoch")},
     {RPMTAG_EPOCH,		0, 0, LEN_AND_STR("serial")},
     {RPMTAG_SUMMARY,		1, 0, LEN_AND_STR("summary")},
+    {RPMTAG_LICENSE,		0, 0, LEN_AND_STR("copyright")},
     {RPMTAG_LICENSE,		0, 0, LEN_AND_STR("license")},
     {RPMTAG_DISTRIBUTION,	0, 0, LEN_AND_STR("distribution")},
     {RPMTAG_DISTURL,		0, 0, LEN_AND_STR("disturl")},
-- 
1.6.4.4

