From 23d1183e6c49d34f7a843b48f92b8b405ced906b Mon Sep 17 00:00:00 2001
From: unknown author <cooker@mandrivalinux.org>
Date: Mon, 5 Jan 2009 13:29:57 +0000
Subject: [PATCH 07/36] ordererase

---
 lib/rpminstall.c |    2 ++
 1 files changed, 2 insertions(+), 0 deletions(-)

diff --git a/lib/rpminstall.c b/lib/rpminstall.c
index 6b3dde7..fcc4eb9 100644
--- a/lib/rpminstall.c
+++ b/lib/rpminstall.c
@@ -259,12 +259,14 @@ static int rpmcliTransaction(rpmts ts, struct rpmInstallArguments_s * ia,
 	ps = rpmpsFree(ps);
     }
 
+#if 0
     if ((eflags? 1 : (!stop)) && !(ia->installInterfaceFlags & INSTALL_NOORDER)) {
 	if (rpmtsOrder(ts)) {
 	    rc = numPackages;
 	    stop = 1;
 	}
     }
+#endif
 
     if (numPackages && !stop) {
 
-- 
1.6.4.4

