From 3f8e8a4963eb0adfa38e42f71dca0c25beafa592 Mon Sep 17 00:00:00 2001
From: unknown author <cooker@mandrivalinux.org>
Date: Mon, 5 Jan 2009 13:29:57 +0000
Subject: [PATCH 03/36] syslog

---
 configure.ac |    6 ++++++
 lib/psm.c    |   21 +++++++++++++++++++++
 system.h     |    4 ++++
 3 files changed, 31 insertions(+), 0 deletions(-)

diff --git a/configure.ac b/configure.ac
index 8c2ef50..4ccaa15 100644
--- a/configure.ac
+++ b/configure.ac
@@ -232,6 +232,12 @@ AC_SUBST(WITH_LZMA_LIB)
 #=================
 
 dnl
+dnl Checking syslog
+dnl
+
+AC_CHECK_HEADERS(syslog.h)
+
+dnl
 dnl Check for features
 dnl
 
diff --git a/lib/psm.c b/lib/psm.c
index 3a9dd70..8eb7f9d 100644
--- a/lib/psm.c
+++ b/lib/psm.c
@@ -1580,6 +1580,16 @@ rpmRC rpmpsmStage(rpmpsm psm, pkgStage stage)
 				NULL, NULL);
 	(void) rpmswExit(rpmtsOp(ts, RPMTS_OP_DBADD), 0);
 
+#if HAVE_SYSLOG_H
+        {
+	  char *s;
+
+	  s = headerFormat(fi->h, "%{NAME}-%{VERSION}-%{RELEASE}", NULL);
+	  syslog(LOG_NOTICE, "[RPM] %s installed\n", s);
+	  s = _free(s);
+	 }
+#endif
+
 	if (rc == RPMRC_OK)
 	    rpmteSetDBInstance(psm->te, headerGetInstance(h));
 	headerFree(h);
@@ -1591,6 +1601,17 @@ rpmRC rpmpsmStage(rpmpsm psm, pkgStage stage)
 	rc = rpmdbRemove(rpmtsGetRdb(ts), rpmtsGetTid(ts),
 				rpmteDBInstance(psm->te), NULL, NULL);
 	(void) rpmswExit(rpmtsOp(ts, RPMTS_OP_DBREMOVE), 0);
+
+#if HAVE_SYSLOG_H
+        {
+	  char *s;
+
+	  s = headerFormat(fi->h, "%{NAME}-%{VERSION}-%{RELEASE}", NULL);
+	  syslog(LOG_NOTICE, "[RPM] %s removed\n", s);
+	  s = _free(s);
+	}
+#endif
+
 	if (rc == RPMRC_OK)
 	    rpmteSetDBInstance(psm->te, 0);
 	break;
diff --git a/system.h b/system.h
index dc053ee..9fd16b4 100644
--- a/system.h
+++ b/system.h
@@ -9,6 +9,10 @@
 #include "config.h"
 #endif
 
+#if HAVE_SYSLOG_H
+#include <syslog.h>
+#endif
+
 #include <sys/types.h>
 
 #include <sys/stat.h>
-- 
1.6.4.4

