Index: rpm-4.10.0/tests/rpmbuild.at
===================================================================
--- rpm-4.10.0.orig/tests/rpmbuild.at
+++ rpm-4.10.0/tests/rpmbuild.at
@@ -133,6 +133,11 @@ RPMDB_INIT
 rm -rf ${TOPDIR}
 
 runroot rpmbuild \
+  -bb --quiet /data/SPECS/sh.spec
+
+runroot rpm -i "${TOPDIR}"/RPMS/noarch/sh-1.0-1.noarch.rpm
+
+runroot rpmbuild \
   -bb --quiet /data/SPECS/hlinktest.spec
 
 runroot rpm -i "${TOPDIR}"/RPMS/noarch/hlinktest-1.0-1.noarch.rpm
Index: rpm-4.10.0/tests/data/SPECS/sh.spec
===================================================================
--- /dev/null
+++ rpm-4.10.0/tests/data/SPECS/sh.spec
@@ -0,0 +1,26 @@
+%define debug_package_and_restore %{nil}
+
+Summary: sh -- dummy sh package
+Name: sh
+Version: 1.0
+Release: 1
+Group: Utilities
+License: GPL
+Distribution: RPM test suite.
+Vendor: Red Hat Software
+Packager: Red Hat Software <bugs@redhat.com>
+URL: http://www.redhat.com
+Buildarch: noarch
+Provides: /bin/sh
+
+%description
+dummy sh package
+
+%clean
+
+%files
+%defattr(-,root,root)
+
+%changelog
+* Tue Oct 20 1998 Jeff Johnson <jbj@redhat.com>
+- create.
Index: rpm-4.10.0/tests/atlocal.in
===================================================================
--- rpm-4.10.0.orig/tests/atlocal.in
+++ rpm-4.10.0/tests/atlocal.in
@@ -24,11 +24,21 @@ RPM_XFAIL=${RPM_XFAIL-1}
 
 function run()
 {
+    TOPDIR="${RPMTEST}/build"
+    RPM_CONFIGDIR="${RPMTEST}/@RPMCONFIGDIR@"
+    RPM_POPTEXEC_PATH="${RPMTEST}/@usrbindir@"
+    HOME="${RPMTEST}"
+    export TOPDIR RPM_CONFIGDIR HOME RPM_POPTEXEC_PATH
     "$@" --define "_topdir ${TOPDIR}" --dbpath="${RPMTEST}/var/lib/rpm/"
 }
 
 function runroot()
 {
+    TOPDIR="/build"
+    RPM_CONFIGDIR="/@RPMCONFIGDIR@"
+    RPM_POPTEXEC_PATH="/@usrbindir@"
+    HOME="/"
+    export TOPDIR RPM_CONFIGDIR HOME RPM_POPTEXEC_PATH
     (cd ${RPMTEST} && \
      MAGIC="/magic/magic" FAKECHROOT_BASE="${RPMTEST}" fakechroot "$@" --define "_topdir /build"
     )
