diff -p -up ./tests/Makefile.am.tv ./tests/Makefile.am
--- ./tests/Makefile.am.tv	2012-04-16 14:35:01.000000000 +0200
+++ ./tests/Makefile.am	2012-06-20 22:02:04.963465011 +0200
@@ -108,7 +108,7 @@ testing$(bindir)/rpmbuild: ../rpmbuild
 	ln -s ../$(bindir) testing/usr/bin
 	for prog in gzip cat patch tar sh ln chmod rm mkdir uname grep sed find file mktemp cut sort diff; do p=`which $${prog}`; ln -s $${p} testing/$${p}; done
 	for d in /proc /sys /selinux /etc/selinux; do if [ -d $${d} ]; then ln -s $${d} testing/$${d}; fi; done
-	(cd testing/magic && file -C)
+	cp /usr/share/misc/magic.mgc testing/magic/ || (cd testing/magic && file -C)
 
 check_DATA = atconfig atlocal $(TESTSUITE)
 check_DATA += testing$(bindir)/rpmbuild
diff -p -up ./tests/Makefile.in.tv ./tests/Makefile.in
--- ./tests/Makefile.in.tv	2012-05-22 12:55:36.000000000 +0200
+++ ./tests/Makefile.in	2012-06-20 22:02:12.763058601 +0200
@@ -587,7 +587,7 @@ testing$(bindir)/rpmbuild: ../rpmbuild
 	ln -s ../$(bindir) testing/usr/bin
 	for prog in gzip cat patch tar sh ln chmod rm mkdir uname grep sed find file mktemp cut sort diff; do p=`which $${prog}`; ln -s $${p} testing/$${p}; done
 	for d in /proc /sys /selinux /etc/selinux; do if [ -d $${d} ]; then ln -s $${d} testing/$${d}; fi; done
-	(cd testing/magic && file -C)
+	cp /usr/share/misc/magic.mgc testing/magic/ || (cd testing/magic && file -C)
 
 @HAVE_FAKECHROOT_TRUE@check-local: $(check_DATA)
 @HAVE_FAKECHROOT_TRUE@	$(SHELL) '$(TESTSUITE)' $(TESTSUITEFLAGS) ||:
diff -p -up ./tests/data/SPECS/hello.spec.tv ./tests/data/SPECS/hello.spec
--- ./tests/data/SPECS/hello.spec.tv	2012-06-26 08:41:23.444615795 +0000
+++ ./tests/data/SPECS/hello.spec	2012-06-26 08:41:17.242604248 +0000
@@ -1,3 +1,5 @@
+%define debug_package_and_restore %{nil}
+
 Summary: hello -- hello, world rpm
 Name: hello
 Version: 1.0
