From 62e44e8f9e9491ea78f1de57619ec2a615fb601b Mon Sep 17 00:00:00 2001
From: unknown author <cooker@mandrivalinux.org>
Date: Mon, 5 Jan 2009 13:29:57 +0000
Subject: [PATCH 04/36] provides obsoleted

---
 lib/depends.c |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/lib/depends.c b/lib/depends.c
index ef55adb..41aa062 100644
--- a/lib/depends.c
+++ b/lib/depends.c
@@ -316,7 +316,7 @@ addheader:
 	     * If no obsoletes version info is available, match all names.
 	     */
 	    if (rpmdsEVR(obsoletes) == NULL
-	     || rpmdsAnyMatchesDep(oh, obsoletes, _rpmds_nopromote)) {
+	     || rpmdsNVRMatchesDep(oh, obsoletes, _rpmds_nopromote)) {
 		char * ohNEVRA = headerGetAsString(oh, RPMTAG_NEVRA);
 #ifdef	DYING	/* XXX see http://bugzilla.redhat.com #134497 */
 		if (rpmVersionCompare(h, oh))
-- 
1.6.4.4

