From 69ed95eb422c59f9d99f75ccb78f32311a31702d Mon Sep 17 00:00:00 2001
From: Lubos Kardos <lkardos@redhat.com>
Date: Fri, 15 Apr 2016 10:45:47 +0200
Subject: [PATCH 2/2] Enable --no-backup-if-mismatch by default in %patch macro
 (rhbz:884755)

---
 macros.in | 1 +
 1 file changed, 1 insertion(+)

diff --git a/macros.in b/macros.in
index bbff0c1..f9172d0 100644
--- a/macros.in
+++ b/macros.in
@@ -364,6 +364,7 @@ package or when debugging this package.\
 
 #	Default patch flags
 #%_default_patch_flags	-s
+%_default_patch_flags --no-backup-if-mismatch
 
 #==============================================================================
 # ---- Build configuration macros.
-- 
1.9.3

