diff -p -up rpm-4.6.1/installplatform.3003 rpm-4.6.1/installplatform
--- rpm-4.6.1/installplatform.3003	2010-09-29 07:36:31.000000000 +0200
+++ rpm-4.6.1/installplatform	2010-09-29 07:37:08.320308967 +0200
@@ -115,7 +115,15 @@ for SUBST in $SUBSTS ; do
 	ISABITS=32
 	;;
     arm*)
-	ISANAME=`echo ${ARCH} | sed "s/^\([^-]*\)-.*/\1/"`
+	ISANAME=armeb
+	ISABITS=32
+	;;
+    arm)
+	ISANAME=armel
+	ISABITS=32
+	;;
+    arm*l)
+	ISANAME=armel
 	ISABITS=32
 	;;
     alpha*)
