--- rpm-4.6.0/lib/psm.c~	2010-05-06 14:36:50.000000000 +0200
+++ rpm-4.6.0/lib/psm.c	2010-05-06 15:10:06.000000000 +0200
@@ -510,7 +510,11 @@
 	rootFd = open(".", O_RDONLY, 0);
 	if (rootFd >= 0) {
 	    if (rootDir != NULL && strcmp(rootDir, "/") && *rootDir == '/')
-		xx = chroot(rootDir);
+		if(chroot(rootDir) < 0) {
+		    rpmlog(RPMLOG_ERR, _("Unable to change root directory: %m\n"));
+		    free(sname);
+		    return rc;
+		}
 	    xx = rpmtsSetChrootDone(ts, 1);
 	}
     }
