Index: rpm-4.8.0/installplatform
===================================================================
--- rpm-4.8.0/installplatform
+++ rpm-4.8.0/installplatform	2010-03-01 16:08:55.495440879 +0100
@@ -98,10 +98,18 @@
 	ISANAME=sh
 	ISABITS=32
 	;;
-    arm*)
-	ISANAME=`echo ${ARCH} | sed "s/^\([^-]*\)-.*/\1/"`
+    arm*b)
+	ISANAME=armeb
 	ISABITS=32
 	;;
+    arm)
+	ISANAME=armel
+	ISABITS=32
+	;;
+    arm*l)
+	ISANAME=armel
+ 	ISABITS=32
+	;;
     alpha*)
 	ISANAME=alpha
 	ISABITS=64
