diff --git a/macros.in b/macros.in
index 5e0d17b..1b5ce0a 100644
--- a/macros.in
+++ b/macros.in
@@ -854,7 +854,7 @@ package or when debugging this package.\
 # because the rest of the arguments will be expanded using %*.
 #
 %_configure ./configure
-%configure \
+%old_configure \
   CFLAGS="${CFLAGS:-%optflags}" ; export CFLAGS ; \
   CXXFLAGS="${CXXFLAGS:-%optflags}" ; export CXXFLAGS ; \
   FFLAGS="${FFLAGS:-%optflags}" ; export FFLAGS ; \
